File: native-plugins/digital-ink/ios/DigitalInkPlugin.swift
// DigitalInkPlugin — Capacitor bridge to ML Kit Digital Ink Recognition (iOS).
//
// STAGED, NOT YET IN THE BUILD. To activate: add this file + DigitalInkPlugin.m
// to the App target, resolve the ML Kit dependency, and flip
// consts::DRAW_MLKIT_READY. See ../README.md for the exact steps.
//
// Signatures verified against the current ML Kit iOS Digital Ink docs
// (developers.google.com/ml-kit/vision/digital-ink-recognition/ios). Per the
// VERIFY-FIRST rule, re-confirm them against the docs at integration time — ML
// Kit APIs shift between releases.
//
// NOTE: iOS ML Kit candidates expose `.text` but NO numeric score (unlike
// Android). We synthesize a rank-derived score (1.0 for rank 0, descending) so
// the core scorer's top-N logic is uniform across platforms.

import Foundation
import Capacitor
import MLKitDigitalInkRecognition

@objc(DigitalInkPlugin)
public class DigitalInkPlugin: CAPPlugin {

    private let workQueue = DispatchQueue(label: "net.spellgame.digitalink", qos: .userInitiated)

    // MARK: - Model management

    private func model(forTag tag: String) -> DigitalInkRecognitionModel? {
        // Full BCP-47 tags only: "zh-Hani-CN", "ja-JP", "ko-KR", "th-TH".
        guard let identifier = DigitalInkRecognitionModelIdentifier(forLanguageTag: tag) else {
            return nil
        }
        return DigitalInkRecognitionModel(modelIdentifier: identifier)
    }

    @objc func downloadModel(_ call: CAPPluginCall) {
        guard let tag = call.getString("languageTag"), let model = model(forTag: tag) else {
            call.reject("unknown or unsupported languageTag", "BAD_TAG")
            return
        }
        let manager = ModelManager.modelManager()
        if manager.isModelDownloaded(model) {
            call.resolve(["status": "already"])
            return
        }
        // Tens of MB; allow cellular per the user's network (Capacitor call is
        // user-initiated on first language selection).
        let conditions = ModelDownloadConditions(allowsCellularAccess: true, allowsBackgroundDownloading: true)
        let progress = manager.download(model, conditions: conditions)
        // Poll the Progress to completion off the main thread; resolve once.
        workQueue.async {
            while !progress.isFinished && !progress.isCancelled {
                Thread.sleep(forTimeInterval: 0.2)
            }
            if progress.isCancelled {
                call.reject("model download cancelled", "DOWNLOAD_CANCELLED")
            } else if manager.isModelDownloaded(model) {
                call.resolve(["status": "downloaded"])
            } else {
                call.reject("model download failed", "DOWNLOAD_FAILED")
            }
        }
    }

    @objc func isModelDownloaded(_ call: CAPPluginCall) {
        guard let tag = call.getString("languageTag"), let model = model(forTag: tag) else {
            call.reject("unknown or unsupported languageTag", "BAD_TAG")
            return
        }
        call.resolve(["downloaded": ModelManager.modelManager().isModelDownloaded(model)])
    }

    @objc func deleteModel(_ call: CAPPluginCall) {
        guard let tag = call.getString("languageTag"), let model = model(forTag: tag) else {
            call.reject("unknown or unsupported languageTag", "BAD_TAG")
            return
        }
        ModelManager.modelManager().deleteDownloadedModel(model) { error in
            if let error = error {
                call.reject("delete failed: \(error.localizedDescription)", "DELETE_FAILED")
            } else {
                call.resolve()
            }
        }
    }

    // MARK: - Recognition

    @objc func recognize(_ call: CAPPluginCall) {
        guard let tag = call.getString("languageTag"), let model = model(forTag: tag) else {
            call.reject("unknown or unsupported languageTag", "BAD_TAG")
            return
        }
        let manager = ModelManager.modelManager()
        guard manager.isModelDownloaded(model) else {
            call.reject("model not downloaded", "MODEL_UNAVAILABLE")
            return
        }
        // strokes: [{ points: [{ x, y, t }] }] — CSS-pixel space, t in ms.
        guard let strokesRaw = call.getArray("strokes") as? [[String: Any]] else {
            call.reject("missing strokes", "BAD_INPUT")
            return
        }
        let width = call.getInt("writingArea.width") ?? (call.getObject("writingArea")?["width"] as? Int ?? 0)
        let height = call.getInt("writingArea.height") ?? (call.getObject("writingArea")?["height"] as? Int ?? 0)
        let preContext = call.getString("preContext") ?? ""

        var strokes: [Stroke] = []
        for strokeDict in strokesRaw {
            guard let pts = strokeDict["points"] as? [[String: Any]] else { continue }
            var points: [StrokePoint] = []
            for p in pts {
                let x = Float((p["x"] as? Double) ?? (p["x"] as? Int).map(Double.init) ?? 0)
                let y = Float((p["y"] as? Double) ?? (p["y"] as? Int).map(Double.init) ?? 0)
                let t = Int((p["t"] as? Int) ?? (p["t"] as? Double).map(Int.init) ?? 0)
                points.append(StrokePoint(x: x, y: y, t: t))
            }
            if !points.isEmpty { strokes.append(Stroke(points: points)) }
        }
        guard !strokes.isEmpty else {
            call.reject("no strokes", "BAD_INPUT")
            return
        }
        let ink = Ink(strokes: strokes)

        let options = DigitalInkRecognizerOptions(model: model)
        let recognizer = DigitalInkRecognizer.digitalInkRecognizer(options: options)
        let context = DigitalInkRecognitionContext(
            preContext: preContext,
            writingArea: WritingArea(width: width, height: height)
        )

        workQueue.async {
            recognizer.recognizeHandwriting(from: ink, context: context) { result, error in
                if let error = error {
                    call.reject("recognize failed: \(error.localizedDescription)", "RECOGNIZE_FAILED")
                    return
                }
                let cands = result?.candidates ?? []
                // iOS gives no score → synthesize a descending rank score.
                let n = max(cands.count, 1)
                let mapped: [[String: Any]] = cands.enumerated().map { (i, c) in
                    ["text": c.text, "score": Double(n - i) / Double(n)]
                }
                call.resolve(["candidates": mapped])
            }
        }
    }
}

File: native-plugins/digital-ink/android/DigitalInkPlugin.kt
// DigitalInkPlugin — Capacitor bridge to ML Kit Digital Ink Recognition (Android).
//
// STAGED, NOT YET IN THE BUILD. iOS is the launch target (spec §4); this keeps
// the interface identical for a future Android build. Add to the app source set
// + the ML Kit gradle dependency, and register in MainActivity (see README.md).
//
// Signatures are from the Android ML Kit Digital Ink API; RE-VERIFY against the
// current docs at integration time (VERIFY-FIRST rule). Unlike iOS, Android
// candidates DO expose a numeric score (candidate.score).

package net.spellgame.app

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.common.model.RemoteModelManager
import com.google.mlkit.vision.digitalink.Ink
import com.google.mlkit.vision.digitalink.DigitalInkRecognition
import com.google.mlkit.vision.digitalink.DigitalInkRecognitionModel
import com.google.mlkit.vision.digitalink.DigitalInkRecognitionModelIdentifier
import com.google.mlkit.vision.digitalink.DigitalInkRecognizerOptions
import com.google.mlkit.vision.digitalink.RecognitionContext
import com.google.mlkit.vision.digitalink.WritingArea

@CapacitorPlugin(name = "DigitalInk")
class DigitalInkPlugin : Plugin() {

    private val remoteModelManager = RemoteModelManager.getInstance()

    private fun modelFor(tag: String): DigitalInkRecognitionModel? {
        val identifier = try {
            DigitalInkRecognitionModelIdentifier.fromLanguageTag(tag)
        } catch (e: Exception) {
            null
        } ?: return null
        return DigitalInkRecognitionModel.builder(identifier).build()
    }

    @PluginMethod
    fun downloadModel(call: PluginCall) {
        val model = modelFor(call.getString("languageTag") ?: "")
            ?: return call.reject("unknown or unsupported languageTag", "BAD_TAG")
        remoteModelManager.isModelDownloaded(model).addOnSuccessListener { already ->
            if (already) {
                call.resolve(JSObject().put("status", "already"))
            } else {
                remoteModelManager
                    .download(model, DownloadConditions.Builder().build())
                    .addOnSuccessListener { call.resolve(JSObject().put("status", "downloaded")) }
                    .addOnFailureListener { e -> call.reject("download failed: ${e.message}", "DOWNLOAD_FAILED") }
            }
        }
    }

    @PluginMethod
    fun isModelDownloaded(call: PluginCall) {
        val model = modelFor(call.getString("languageTag") ?: "")
            ?: return call.reject("unknown or unsupported languageTag", "BAD_TAG")
        remoteModelManager.isModelDownloaded(model)
            .addOnSuccessListener { call.resolve(JSObject().put("downloaded", it)) }
            .addOnFailureListener { call.resolve(JSObject().put("downloaded", false)) }
    }

    @PluginMethod
    fun deleteModel(call: PluginCall) {
        val model = modelFor(call.getString("languageTag") ?: "")
            ?: return call.reject("unknown or unsupported languageTag", "BAD_TAG")
        remoteModelManager.deleteDownloadedModel(model)
            .addOnSuccessListener { call.resolve() }
            .addOnFailureListener { e -> call.reject("delete failed: ${e.message}", "DELETE_FAILED") }
    }

    @PluginMethod
    fun recognize(call: PluginCall) {
        val model = modelFor(call.getString("languageTag") ?: "")
            ?: return call.reject("unknown or unsupported languageTag", "BAD_TAG")
        val strokesArr = call.getArray("strokes")
            ?: return call.reject("missing strokes", "BAD_INPUT")
        val area = call.getObject("writingArea")
        val preContext = call.getString("preContext") ?: ""

        val inkBuilder = Ink.builder()
        for (i in 0 until strokesArr.length()) {
            val strokeObj = strokesArr.getJSONObject(i)
            val pts = strokeObj.getJSONArray("points")
            val strokeBuilder = Ink.Stroke.builder()
            for (j in 0 until pts.length()) {
                val p = pts.getJSONObject(j)
                strokeBuilder.addPoint(
                    Ink.Point.create(
                        p.getDouble("x").toFloat(),
                        p.getDouble("y").toFloat(),
                        p.optLong("t", 0L),
                    )
                )
            }
            inkBuilder.addStroke(strokeBuilder.build())
        }
        val ink = inkBuilder.build()

        val recognizer = DigitalInkRecognition.getClient(
            DigitalInkRecognizerOptions.builder(model).build()
        )
        val context = RecognitionContext.builder()
            .setPreContext(preContext)
            .setWritingArea(
                WritingArea(
                    (area?.getInt("width") ?: 0).toFloat(),
                    (area?.getInt("height") ?: 0).toFloat(),
                )
            )
            .build()

        recognizer.recognize(ink, context)
            .addOnSuccessListener { result ->
                val out = JSObject()
                val arr = com.getcapacitor.JSArray()
                for (c in result.candidates) {
                    arr.put(JSObject().put("text", c.text).put("score", c.score ?: 0.0))
                }
                out.put("candidates", arr)
                call.resolve(out)
            }
            .addOnFailureListener { e -> call.reject("recognize failed: ${e.message}", "RECOGNIZE_FAILED") }
    }
}
